### PR TITLE
Cleanup CompiledScriptBlock.cs

### PR DESCRIPTION
**Pull request by vexx32**
# PR Summary

- Refactor for newer & tidier code patterns
- Avoid super long lines where possible
- Consolidate style and patterns

Shouldn't be any functional changes. 

## PR Context

More code cleanup! ^^

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 1]
```diff
@@ -1433,11 +1463,16 @@ private static bool WriteScriptBlockToLog(ScriptBlock scriptBlock, int segment,
                     // version.
                     if (s_encryptionRecipients != null)
                     {
-                        // Encrypt the raw Text from the scriptblock.  The user may have to deal with any control characters in the data.
+                        // Encrypt the raw Text from the scriptblock.  The user may have to deal with any control
+                        // characters in the data.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: It is better split by sentence.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 2]
```diff
@@ -1448,9 +1483,20 @@ private static bool WriteScriptBlockToLog(ScriptBlock scriptBlock, int segment,
...
+                                SecuritySupportStrings.CouldNotEncryptContent,
+                                textToLog,
+                                error.ToString());
+                            PSEtwLog.LogOperationalError(
+                                id: PSEventId.ScriptBlock_Compile_Detail,
+                                opcode: PSOpcode.Create,
+                                task: PSTask.ExecuteCommand,
+                                keyword: PSKeyword.UseAlwaysOperational,
+                                0,
+                                0,
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Please use named parameters.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 3]
```diff
@@ -1542,9 +1606,19 @@ private static bool GetAndValidateEncryptionRecipients(ScriptBlock scriptBlock,
...
+                        string errorMessage = StringUtil.Format(
+                            SecuritySupportStrings.CouldNotUseCertificate,
+                            error.ToString());
+                        PSEtwLog.LogOperationalError(
+                            id: PSEventId.ScriptBlock_Compile_Detail,
+                            opcode: PSOpcode.Create,
+                            task: PSTask.ExecuteCommand,
+                            keyword: PSKeyword.UseAlwaysOperational,
+                            0,
+                            0,
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: The same.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 4]
```diff
@@ -1567,9 +1641,19 @@ private static bool GetAndValidateEncryptionRecipients(ScriptBlock scriptBlock,
...
+                            string errorMessage = StringUtil.Format(
+                                SecuritySupportStrings.CertificateContainsPrivateKey,
+                                certificateForLog);
+                            PSEtwLog.LogOperationalError(
+                                id: PSEventId.ScriptBlock_Compile_Detail,
+                                opcode: PSOpcode.Create,
+                                task: PSTask.ExecuteCommand,
+                                keyword: PSKeyword.UseAlwaysOperational,
+                                0,
+                                0,
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: The same.

**Review by iSazonov — COMMENTED** [review]
Thanks for the style fixes!

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 2 above)
Comment (vexx32) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Those last two are `params args`; not sure if applying names to a `params` argument actually works? I'll see if omnisharp complains.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 5]
```diff
@@ -684,29 +700,39 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
...
+            bool filterNonUsingVariables,
+            bool? createLocalScope,
+            params object[] args)
+            => AstInternal.GetPowerShell(
+                context,
+                variables,
+                isTrustedInput,
+                filterNonUsingVariables,
+                createLocalScope,
+                args);
```
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Hmm. Not sure that'll make it look much better that way either, just so many arguments everywhere. I'll see if that helps, but darned if c# doesn't need @psboundparameters eh? :smile:

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 6]
```diff
@@ -554,13 +577,13 @@ internal static ScriptBlock TryGetCachedScriptBlock(string fileName, string file
         private static bool IsDynamicKeyword(Ast ast)
         {
             var cmdAst = ast as CommandAst;
-            return (cmdAst != null && cmdAst.DefiningKeyword != null);
+            return cmdAst?.DefiningKeyword != null;
```
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: I'm almost certain this kind of pattern should give the correct result, this is a textbook usage of `?.

The other one... Yeah, might be a bit funky.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Ok, ignore my request.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 7]
```diff
@@ -554,13 +577,13 @@ internal static ScriptBlock TryGetCachedScriptBlock(string fileName, string file
...
         }
 
         private static bool IsUsingTypes(Ast ast)
         {
             var cmdAst = ast as UsingStatementAst;
-            return (cmdAst != null && cmdAst.IsUsingModuleOrAssembly());
+            return cmdAst?.IsUsingModuleOrAssembly() == true;
```
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Verified this in a simple test case with SharpLab. As long as the method returns a bool value, the compiler sees basically the same thing. Current code vs the original compiled & decompiled to show what this is essentially equivalent to:

image

[SharpLab Link](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBEJinANjASQDsYIFsBjCAgWwAdcIaITYBLAeyIB8ABAJgCMAWABQvAMwACAVICqyNkQDmAZRiwmLGAEFkMKSCl6DAbzFTLM6bwAsUgLIAKAJRTzoq1IC+Fq38tJKWAODlxjAG4A62DQ8IJkBSVlBw4AEzwIAHkEPWQmYFwAT1d3GQB2YykAfikYBDgIQylsXHyInzFfUTEg2RN3aLoENgA3TViwqSIOOiaAXhaoNogozythsYmQqdUICFLTCunZ1c7Rbt6bfikAFRQzIZHx0hlBADZJ+MTFFVuiubIJwDKD6NweLxecYIKQUGhpAaLUEGUHyX5qDSkZgkEwdSGWaJeXiVOEI/TVAB0CSSKlSGUYOTyBWKpXmi3qjTWXm6XieW1evA+XykOTYyiUy2p6P+gOB+ikyPBhKs0Nh8MRCvlqJpGM02N0+jx+OVgRJ6vlAEJFkQ8OEAGR2tVkmBUn7JOmZRnIfI0QolFxcqzdbxAA) to fiddle with if you feel the need. 🙂

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 8]
```diff
@@ -1100,7 +1123,10 @@ internal bool IsUsingDollarInput()
                             }
 
                             string name = psvar.Name;
-                            Diagnostics.Assert(!(string.Equals(name, "this") || string.Equals(name, "_") || string.Equals(name, "input")),
+                            Diagnostics.Assert(
+                                !(string.Equals(name, "this")
+                                    || string.Equals(name, "_")
+                                    || string.Equals(name, "input")),
```
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: That's true. I almost want to say we should string.Format() that but that's excessive... probably. :)

**Review by iSazonov — APPROVED** [review]

**Comment by vexx32** [comment]
@iSazonov did you guys need anything more from me here, or is this good to be merged? 🙂

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: I still think we need revert.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: I don't see benefits in style. I still think we need revert.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 9]
```diff
@@ -890,76 +915,72 @@ internal string GetWithInputHandlingForInvokeCommand()
...
+                createLocalScope,
+                functionsToDefine,
+                variablesToDefine,
+                errorHandlingBehavior,
+                dollarUnder,
+                input,
+                scriptThis,
+                outputPipe,
+                invocationInfo,
+                args);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: I don't see benefits in the style. Please revert.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 10]
```diff
@@ -1611,12 +1693,13 @@ internal static string CheckSuspiciousContent(Ast scriptBlockAst)
...
-                {
-                    // Try to find the lowest AST that was not considered suspicious, but its parent
-                    // was.
-                    return (!ast.HasSuspiciousContent) && (ast.Parent.HasSuspiciousContent);
-                }, true);
+                Ast foundAst = scriptBlockAst.Find(
+                    ast =>
+                    {
+                        // Try to find the lowest AST that was not considered suspicious, but its parent
+                        // was.
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Can we make one line comment?

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (vexx32) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Fair enough!

I reverted it and the code analysis indicated we could use this pattern instead. What do you think?

```csharp
return ast is CommandAst cmdAst && cmdAst.DefiningKeyword != null;
```

[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Looks good.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 11]
```diff
@@ -570,25 +587,23 @@ internal static void CacheScriptBlock(ScriptBlock scriptBlock, string fileName,
...
+            if (scriptBlock.Ast.Find(ast => IsUsingTypes(ast), false) != null
+                || scriptBlock.Ast.Find(ast => IsDynamicKeyword(ast), true) != null)
             {
                 return;
             }
 
             if (s_cachedScripts.Count > 1024)
             {
-                s_cachedScripts.Clear();
+                ClearScriptBlockCache();
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: This reduces readability - in lines 604 and 618 we use `s_cachedScripts`. Please revert.

Comment (vexx32) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Should I remove the `ClearScriptBlockCache()` method, then?

If I remove this reference it isn't being used anywhere at all, and it's `internal`.

**Comment by TravisEz13** [comment]
@PoshChan Please rerun windows

**Comment by PoshChan** [comment]
@TravisEz13, successfully started rebuild of `PowerShell-CI-Windows`

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 12]
```diff
@@ -237,10 +246,13 @@ bool IsScriptBlockInFactASafeHashtable()
                 // are enhanced, such as new members added to 'ScriptBlockAst', the code here needs
```
Comment (TravisEz13) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]:
```suggestion
                // are enhanced,  such as new members added to 'ScriptBlockAst', the code here needs
```

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 11 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: It will not a style fix - code cleanup.

**Review by iSazonov — APPROVED** [review]

**Comment by vexx32** [comment]
@iSazonov all green! 🎉 😄

**Comment by TravisEz13** [comment]
@vexx32 Codacy is saying you are introducing more issues than you are fixing.

Can you please review those issues?

**Comment by vexx32** [comment]
@TravisEz13 yeah I've seen them. Codacy sees the inline variable declaration pattern `if (myVar is Type newVar)` and thinks I'm putting in an if statement that lacks braces. Something in its logic is simply incorrect, not sure exactly what's up, but it's not handling these right:

image

It's also making recommendations about removing methods / changing method access modifiers, which I think is probably better determined by you guys if we need to action on any of those recommendations than me tinkering around. 😄 

There are a few unused methods/props here too, unsure if you want anything done with those.

**Comment by iSazonov** [comment]
Codacy doesn't know new C# features and reports many false positives.

**Comment by TravisEz13** [comment]
@adityapatwardhan @vexx32 	@iSazonov 	 	should we disable codacy?

**Comment by adityapatwardhan** [comment]
We can maybe disable everything except PowerShell on Codacy.

**Comment by iSazonov** [comment]
Codacy  does useful recommendations too which we follow and which is not in CodeFactor. 
It makes some recommendations unacceptable to us but I find the tool useful as whole. Perhaps we need configure it more carefully.

**Comment by iSazonov** [comment]
We could have two configs for CodeFactor and Codacy.
First minimal set with mandatory rules for mandatory CI (to block PR merging)
Second config for optional CI with extended set rules (like complex methods, XML docs...)

**Comment by vexx32** [comment]
Sounds good to me. I agree that it does make some more insightful suggestions at times. But there are also times where it just... makes no sense. Higher rate of false positives, I guess you could say. 🙂

**Review by TravisEz13 — APPROVED** [review]

**Comment by vexx32** [comment]
@TravisEz13 @anmenaga any chance we could get this one merged soon-ish as well? 💖 

Once this is in, #9900 is ready to go. 😁

**Comment by anmenaga** [comment]
@vexx32 Please fix the merge conflict with CompiledScriptBlock.cs and I'll merge this. Thank you.

**Comment by vexx32** [comment]
@anmenaga That ought to do it, I think. Let me know if there's anything more y'all need. 🙂

**Pull request merged**